test/BerfChatStorage.ts

```diff
@@ -175,12 +169,204 @@ describe("BerfChatStorage contract tests", async () => {
             EthCrypto.hash.keccak256(decryptedPayload.message)
         );
 
-        console.log(
-            'Got message from ' +
-            senderAddress +
-            ': ' +
-            decryptedPayload.message
+        expect(senderAddress).to.equal(accountUno.address);
+
+        expect(decryptedPayload.message).to.equal(secretMessage);
+    })
+
+    it("confirms that only the public key's private key can decrypt the message", async () => {
+        // Create new accounts
+        // (Used the EthCrypto library to create accounts
+        // rather than using Hardhat because for testing
+        // purposes, it is much easier to retrieve 
+        // public keys with it.)
+        const accountUno = EthCrypto.createIdentity();
+        const accountDos = EthCrypto.createIdentity();
+        const accountTres = EthCrypto.createIdentity();
+
+        // Declare and assign variable
+        // with secret string.
+        const secretMessage = "I know who Satoshi is.";
+
+        // Sign the message with accountUno's
+        // private key
+        const signature = EthCrypto.sign(
+            accountUno.privateKey,
+            EthCrypto.hash.keccak256(secretMessage)
         );
-        */
+
+        // Create object with message and signature
+        const payload = {
+            message: secretMessage,
+            signature
+        };
+
+        // Encrypt the message and the signature with
+        // accountDos' publicKey. (By encrypting it
+        // with accountDos' public key, only accountDos
+        // can decrypt `payload` with its private key.)
+        const encrypted = await EthCrypto.encryptWithPublicKey(
+            accountDos.publicKey,
+            JSON.stringify(payload) // Stringify the payload to allow encryption
+        );
+
+        // Convert object into smaller string-representation
+        const encryptedString = EthCrypto.cipher.stringify(encrypted);
+
+
+        // ~~~~ SENDING MESSAGE FROM accountOne TO accountTwo ~~~ //
+
+
+        // Parse the string back to an object
+        const encryptedObject = EthCrypto.cipher.parse(encryptedString);
+
+        // Confirm that acountTres CANNOT decrypt message
+        await rejects(EthCrypto.decryptWithPrivateKey(accountTres.privateKey, encryptedObject));
+    });
+    */
+
+    it("tests sending a message via the contract", async () => {
+        // Declare and assign variable
+        // with secret string.
+        const secretMessage: string = "Satoshi is not Vitalik.";
+
+        const signature: string = EthCrypto.sign(
```

Comment:
Why do you sign the message prior to encrypting with the recipeint public key?

Reply:
The tutorial of the `ETH crypto` library laid it out in this way. From my understanding, the recipient calls `decryptWithPrivateKey`, which takes two parameters: the recipient's private key and the encrypted object.

If the message were to be encrypted with the recipients's public key, *then* signed, the recipient wouldn't have the right object to pass into the `decryptWithPrivateKey` function.

And since they still need the signature to derive the sender's address, the message is signed by the sender first before encrypting with the recipient's public key.

Essentially, it comes down to a design decision by the developers of the library.

Is there a concern you had in mind with how it currently works?

I suppose it could lead to a sort of unintuitive UI, but we would need to mock it up to really see, IMO.